File: Second/SecondProject/Plant-entity classes/Bush.cpp
//------------------------------------------------------------------------------
// Bush.cpp - provides implementation for the bush-struct
// and functions working with it.
//------------------------------------------------------------------------------
#include "Bush.h"

// Default "bush"-object constructor.
Bush::Bush() : Plant() {
    this->month_ = BlossomMonth::January;
}

/* Special constructor, which has different ways of creating a
 * "bush"-object depending on what type of input was selected. */
Bush::Bush(bool from_file, FILE *file) : Plant(from_file, file) {
    if (from_file) {
        int month_id;
        fscanf(file, "%d", &month_id);
        month_id = abs(month_id);
        month_id = ((month_id > 12 || !month_id) ? month_id % 12 + 1 : month_id);
        this->month_ = static_cast<Bush::BlossomMonth>(month_id);
    } else {
        this->month_ = static_cast<Bush::BlossomMonth>(rand() % 12 + 1);
    }
}


// Sets the C-string-type data of a given bush into a char**-pointer.
char *Bush::toString() {
    char *plant_data = static_cast<Plant>(*this).toString();
    char *month_data = this->getMonth();
    char *data = new char[101];
    data[100] = '\0';
    int print_count = snprintf(data, 100, "Bush %s:%s", plant_data, month_data);
    data[print_count] = '\0';
    delete [] plant_data;
    delete [] month_data;
    return data;
}

// Gets a C-string representation of a month of blossom of a specific bush.
char *Bush::getMonth() {
    char *data = new char[21];
    data[20] = '\0';
    int symbols_count;
    switch(this->month_) {
        case Bush::BlossomMonth::January: symbols_count = snprintf(data, 20, "January"); break;
        case Bush::BlossomMonth::February: symbols_count = snprintf(data, 20, "February"); break;
        case Bush::BlossomMonth::March: symbols_count = snprintf(data, 20,"March"); break;
        case Bush::BlossomMonth::April: symbols_count = snprintf(data, 20,"April"); break;
        case Bush::BlossomMonth::May: symbols_count = snprintf(data, 20,"May"); break;
        case Bush::BlossomMonth::June: symbols_count = snprintf(data,20, "June"); break;
        case Bush::BlossomMonth::July: symbols_count = snprintf(data,20, "July"); break;
        case Bush::BlossomMonth::August: symbols_count = snprintf(data, 20,"August"); break;
        case Bush::BlossomMonth::September: symbols_count = snprintf(data,20, "September"); break;
        case Bush::BlossomMonth::October: symbols_count = snprintf(data, 20,"October"); break;
        case Bush::BlossomMonth::November: symbols_count = snprintf(data,20, "November"); break;
        case Bush::BlossomMonth::December: symbols_count = snprintf(data, 20,"December"); break;
    }
    data[symbols_count] = '\0';
    return data;
}

File: Third/ThirdProject/plants/Bush.py
# ------------------------------------------------------------------------------
# Bush.py - provides declaration and implementation for the bush-class
# and functions working with it.
# ------------------------------------------------------------------------------
import random
from plants.Plant import Plant
from enum import IntEnum


class Bush(Plant):
    class BlossomMonth(IntEnum):
        JANUARY = 1
        FEBRUARY = 2
        MARCH = 3
        APRIL = 4
        MAY = 5
        JUNE = 6
        JULY = 7
        AUGUST = 8
        SEPTEMBER = 9
        OCTOBER = 10
        NOVEMBER = 11
        DECEMBER = 12

    def __init__(self, is_random: bool, params: list) -> None:
        """
        Special constructor, which has different ways of creating a
        "bush"-object depending on what type of input was selected.

        :param is_random: Logical parameter, specifying a method of extracting input data -
        random generation or file-input.

        :param params: A list of params passed in case of file-input.
        """
        super().__init__(is_random, params)
        if is_random:
            self.blossom_month_ = Bush.BlossomMonth(random.randint(1, 12))
        else:
            if params.__len__() < 3:
                self = 0
                return
            self.blossom_month_ = abs(int(params[2]))
            self.blossom_month_ = Bush.BlossomMonth(self.blossom_month_ % 12 + 1
                                                    if ((self.blossom_month_ == 0) or (self.blossom_month_ > 12))
                                                    else self.blossom_month_)

    def getMonth(self) -> str:
        """
        A function to get string-type presentation of a month of blossom for a specific bush.

        :return: A string, containing information about a month when a bush blossoms in.
        """
        month_switch = {
            Bush.BlossomMonth.JANUARY: "January",
            Bush.BlossomMonth.FEBRUARY: "February",
            Bush.BlossomMonth.MARCH: "March",
            Bush.BlossomMonth.APRIL: "April",
            Bush.BlossomMonth.MAY: "May",
            Bush.BlossomMonth.JUNE: "June",
            Bush.BlossomMonth.JULY: "July",
            Bush.BlossomMonth.AUGUST: "August",
            Bush.BlossomMonth.SEPTEMBER: "September",
            Bush.BlossomMonth.OCTOBER: "October",
            Bush.BlossomMonth.NOVEMBER: "November",
            Bush.BlossomMonth.DECEMBER: "December"
        }
        return month_switch.get(self.blossom_month_, "undefined")

    def toString(self) -> str:
        """
        A function to get string-type presentation of bush-data.

        :return: A string, containing information about a bush.
        """
        return f"Bush {super().toString()}: {self.getMonth()}"
